prow/plugins/projectmanager/projectmanager.go

```diff
@@ -0,0 +1,204 @@
+							configString = fmt.Sprintf("%s \"%s\"", configString, label)
+							if i+1 < len(managedColumn.Labels) {
+								configString = fmt.Sprintf("%s,", configString)
+							}
+						}
+						configString = fmt.Sprintf("%s ] }", configString)
+					}
+					configString = fmt.Sprintf("%s ] }", configString)
+				}
+				configString = fmt.Sprintf("%s }", configString)
+			}
+			configString = fmt.Sprintf("%s }", configString)
+			configMap[""] = configString
+			return configMap
+		}(config),
+	}
+	return pluginHelp, nil
+}
+
+func handlePullRequest(pc plugins.Agent, pe github.PullRequestEvent) error {
+	return handlePR(pc.GitHubClient, pc.PluginConfig.ProjectManager, pc.Logger, pe)
+}
+
+// Strict subset of *github.Client methods.
+type githubClient interface {
+	GetIssueLabels(org, repo string, number int) ([]github.Label, error)
+	GetRepoProjects(owner, repo string) ([]github.Project, error)
+	GetOrgProjects(org string) ([]github.Project, error)
+	GetProjectColumns(projectID int) ([]github.ProjectColumn, error)
+	CreateProjectCard(columnID int, projectCard github.ProjectCard) (*github.ProjectCard, error)
+}
+
+func handlePR(gc githubClient, projectManager plugins.ProjectManager, log *logrus.Entry, pe github.PullRequestEvent) error {
+	// Only respond to label add or issue/PR open events
+	if pe.Action != github.PullRequestActionOpened &&
+		pe.Action != github.PullRequestActionReopened &&
+		pe.Action != github.PullRequestActionLabeled {
+		return nil
+	}
+	// Get any ManagedProjects that match this PR
+	matchedColumnIDs, err := getMatchingColumnIDs(gc, projectManager, pe)
+	if err != nil {
+		return err
+	}
+	// For each ManagedColumn that matches this PR, add this PR to that Project Column
+	for _, matchedColumnID := range matchedColumnIDs {
+		err = addPRToColumn(gc, matchedColumnID, pe)
+		log.WithError(err).Println("Failed to add PR to project")
+	}
+	return nil
+}
+
+func getMatchingColumnIDs(gc githubClient, projectManager plugins.ProjectManager, pe github.PullRequestEvent) ([]int, error) {
+	var matchedColumnIDs []int
+	// Don't use GetIssueLabels unless it's required and keep track of whether the labels have been fetched to avoid unnecessary API usage.
+	labelsFetched := false
+	var labels []github.Label
+	for orgRepoName, managedOrgRepo := range projectManager.OrgRepos {
+		for projectName, managedProject := range managedOrgRepo.Projects {
+			for _, managedColumn := range managedProject.Columns {
+				if managedColumn.Org != "" && managedColumn.Org != pe.Repo.Owner.Login {
+					continue
+				}
+				if managedColumn.State != "" && managedColumn.State != pe.PullRequest.State {
+					continue
+				}
+				if len(managedColumn.Labels) != 0 {
+					if !labelsFetched {
+						// If labels are not yet fetched then get them as they are now required
+						// GetIssueLabels works for PRs as they are considered issues in the API
+						var err error
+						labels, err = gc.GetIssueLabels(pe.Repo.Owner.Login, pe.Repo.Name, pe.Number)
```

Comment:
IIRC we get the labels on the event? So we do not need to do this fetch?

Reply:
Yes, if we just react to PullRequestEvents with Action 'labeled' which will be triggered for each of the labels. We only have to perform necessary action for that label. Currently code processes 'Open' and other PR Actions and we should ignore those. WDYT